### PR TITLE
[wasm] Switch default modules to es6

### PR DESCRIPTION
**Pull request by pavelsavara**
## Motivation, benefits 
- TL;DR: 
    - consistent in NodeJS and in browsers
    - module isolation
    - well defined exports/imports
    - dependencies resolution
    - dependencies detected during parsing, not during run
    - async dynamic imports awaitable
- [Why Modules](https://eloquentjavascript.net/10_modules.html) [cartoon](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/)
- Supported everywhere for many years, everywhere where WASM is able to run.

## Challenges, downsides
- CommonJS module could be [wrapped inside simple ES6 script](https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1). **But not the other way around.**
    - ES6 modules could be re-packaged by standard tools like webpack, rollup back to CJS or UMD.
    - libraries using another ES6 modules have to be also ES6 module (unless they re-pack with tools above).
- Node.js 12 doesn't support it, but is end-of-life by 2022-04-30
- After this PR will would still have way how to produce CommonJS, but [is it worth supporting it](https://github.com/dotnet/runtime/issues/63337) ?

## Changes in this PR
- switched dotnet.js to be ES6 module by making `<WasmEnableES6>` default true
- updated all samples
- updated functional tests
- updated debugger tests
- updated test-main to
- `EXPORT_ES6=1` should not be in the `emcc-link.rsp` because it could not be removed by workload re-link when `<WasmEnableES6> == false`.

### CODE REVIEW

**Comment by pavelsavara** [comment]
- I fixed and tested samples:
browser\
browser-bench\
browser-cjs\
browser-legacy\
browser-nextjs\
browser-profile\
browser-webpack\
console-node-cjs\
console-node-es6\
console-node-ts\
console-v8-cjs\
console-v8-es6\
node-webpack\
mbr\browser

- unit tests:
src/tests/FunctionalTests/WebAssembly/Browser/HotReload/
src/tests/FunctionalTests/WebAssembly/Browser/RuntimeConfig
src/mono/wasm/debugger/DebuggerTestSuite 

- I tested that it works in Blazor after the change
- I tested that debugger works in Blazor

##TODO
- right now the worker doesn't work for me with /p:WasmEnablePerfTracing=true: 
  - browser-eventpipe\ browser-mt-eventpipe\
  - the dotnet.worker.js has `require` in it
  - the dotnet.worker.js has eval in it
- the console template could become es6 as well
- make sure that all WBT work

**Comment by pavelsavara** [comment]
```
[FAIL] System.Runtime.InteropServices.JavaScript.Tests.TypedArrayTests.Uint8ArrayToArray
[13:50:25] info: System.Runtime.InteropServices.JavaScript.JSException : RangeError: Array buffer allocation failed
[13:50:25] info:     at new ArrayBuffer (<anonymous>)
[13:50:25] info:     at eval (eval at t (http://127.0.0.1:33131/dotnet.js:3:44553), <anonymous>:3:23)
[13:50:25] info:     at Object.mono_wasm_invoke_js_with_args_ref (http://127.0.0.1:33131/dotnet.js:3:59772)
```

**Comment by pavelsavara** [comment]
Errors on `runtime-wasm ` CI 
- `SharedArrayBuffer is not defined` 
- `requested a shared WebAssembly.Memory but the returned buffer is not a SharedArrayBuffer, indicating that while the browser has SharedArrayBuffer it does not have WebAssembly threads support - you may need to set a flag` 

are https://github.com/dotnet/runtime/issues/71196

File: src/mono/wasm/templates/templates/console/console.0.csproj [hunk 1]
```diff
@@ -5,13 +5,13 @@
     <TargetOS>browser</TargetOS>
     <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
     <UseMonoRuntime>true</UseMonoRuntime>
-    <WasmMainJSPath>main.cjs</WasmMainJSPath>
+    <WasmMainJSPath>main.mjs</WasmMainJSPath>
     <OutputType>Exe</OutputType>
-    <WasmEnableES6>false</WasmEnableES6>
+    <WasmEnableES6>true</WasmEnableES6>
```
Comment (maraf) [inline src/mono/wasm/templates/templates/console/console.0.csproj:10]: True is a new default, right?

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs [hunk 2]
```diff
@@ -23,6 +23,8 @@ public class WasmAppBuilder : Task
     [Required]
     public string? MainJS { get; set; }
 
+    public string? WasmEnableES6 { get; set; }
```
Comment (maraf) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:26]: Shouldn't it be a `bool?`?

**Review by ilonatommy — COMMENTED** [review]
I can't find anything wrong. Good job

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs (hunk 2 above)
Comment (pavelsavara) [reply src/tasks/WasmAppBuilder/WasmAppBuilder.cs:26]: Somehow it didn't work well for me and it was not worth fighting it.

**Review by maraf — APPROVED** [review]
Good job 👏

File: src/mono/sample/wasm/console-node-ts/Wasm.Console.Node.TS.Sample.csproj [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
+    We need to install typescript definition (so tsc knows runtime types) and package.json (tsc knows about runtime files) from artifacts directory.
+  -->
+  <Target Name="CopyRelinkedPackage" AfterTargets="WasmBuildApp" DependsOnTargets="Build" Inputs="
+          $(MicrosoftNetCoreAppRuntimePackNativeDir)/dotnet.d.ts;
+          $(MicrosoftNetCoreAppRuntimePackNativeDir)/package.json;" Outputs="bin/$(Configuration)/AppBundle/.npm-stamp">
+    <ItemGroup>
+      <NpmPackageFiles Include="$(MicrosoftNetCoreAppRuntimePackNativeDir)/dotnet.d.ts"/>
+      <NpmPackageFiles Include="$(MicrosoftNetCoreAppRuntimePackNativeDir)/package.json"/>
+    </ItemGroup>
+    <Copy SourceFiles="@(NpmPackageFiles)" DestinationFolder="bin/$(Configuration)/AppBundle" />
```
Comment (radical) [inline src/mono/sample/wasm/console-node-ts/Wasm.Console.Node.TS.Sample.csproj:18]: Add `SkipUnchangedFiles="true"`

File: src/mono/sample/wasm/console-node-ts/Wasm.Console.Node.TS.Sample.csproj [hunk 4]
```diff
@@ -1,15 +1,28 @@
...
-  <Target Name="TypeScript" AfterTargets="WasmBuildApp" DependsOnTargets="Build">
+  <!-- 
+    We need to install typescript definition (so tsc knows runtime types) and package.json (tsc knows about runtime files) from artifacts directory.
+  -->
+  <Target Name="CopyRelinkedPackage" AfterTargets="WasmBuildApp" DependsOnTargets="Build" Inputs="
+          $(MicrosoftNetCoreAppRuntimePackNativeDir)/dotnet.d.ts;
+          $(MicrosoftNetCoreAppRuntimePackNativeDir)/package.json;" Outputs="bin/$(Configuration)/AppBundle/.npm-stamp">
+    <ItemGroup>
+      <NpmPackageFiles Include="$(MicrosoftNetCoreAppRuntimePackNativeDir)/dotnet.d.ts"/>
+      <NpmPackageFiles Include="$(MicrosoftNetCoreAppRuntimePackNativeDir)/package.json"/>
```
Comment (radical) [inline src/mono/sample/wasm/console-node-ts/Wasm.Console.Node.TS.Sample.csproj:16]: nit: You could move this outside the target, and then use the item reference in the `Inputs` too.

File: src/mono/wasm/runtime/CMakeLists.txt [hunk 5]
```diff
@@ -30,9 +30,8 @@ target_link_libraries(dotnet
...
-    RUNTIME_OUTPUT_DIRECTORY "${NATIVE_BIN_DIR}")
+    LINK_DEPENDS "${NATIVE_BIN_DIR}/src/emcc-default.rsp;${NATIVE_BIN_DIR}/src/es6/dotnet.es6.pre.js;${NATIVE_BIN_DIR}/src/es6/runtime.es6.iffe.js;${NATIVE_BIN_DIR}/src/es6/dotnet.es6.lib.js;${NATIVE_BIN_DIR}/src/pal_random.lib.js;${NATIVE_BIN_DIR}/src/es6/dotnet.es6.post.js;"
+    LINK_FLAGS "@${NATIVE_BIN_DIR}/src/emcc-default.rsp @${NATIVE_BIN_DIR}/src/emcc-link.rsp ${CONFIGURATION_LINK_FLAGS} --extern-pre-js ${NATIVE_BIN_DIR}/src/es6/runtime.es6.iffe.js --pre-js ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.pre.js --js-library ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.lib.js --js-library ${NATIVE_BIN_DIR}/src/pal_random.lib.js --post-js ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.post.js "    RUNTIME_OUTPUT_DIRECTORY "${NATIVE_BIN_DIR}")
```
Comment (radical) [inline src/mono/wasm/runtime/CMakeLists.txt:34]: nit:
```suggestion
    LINK_FLAGS "@${NATIVE_BIN_DIR}/src/emcc-default.rsp @${NATIVE_BIN_DIR}/src/emcc-link.rsp ${CONFIGURATION_LINK_FLAGS} --extern-pre-js ${NATIVE_BIN_DIR}/src/es6/runtime.es6.iffe.js --pre-js ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.pre.js --js-library ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.lib.js --js-library ${NATIVE_BIN_DIR}/src/pal_random.lib.js --post-js ${NATIVE_BIN_DIR}/src/es6/dotnet.es6.post.js "
    RUNTIME_OUTPUT_DIRECTORY "${NATIVE_BIN_DIR}")
```

File: src/mono/wasm/templates/templates/browser/app-support.js [hunk 6]
```diff
@@ -149,44 +147,43 @@ function applyArguments() {
...
-                        } else {
-                            runArgs = await response.json();
-                            console.debug(`runArgs: ${JSON.stringify(runArgs)}`);
-                        }
-                    })
-                    .catch(error => console.error(`Failed to load args: ${stringify_as_error_with_stack(error)}`));
-
-Promise.all([ argsPromise, loadDotnetPromise ]).then(async ([ _, { default: createDotnetRuntime } ]) => {
+try {
+    initRunArgs();
```
Comment (radical) [inline src/mono/wasm/templates/templates/browser/app-support.js:163]: We want to load to populate `runArgs` with contents from `runArgs.json`, then `initRunArgs`, and then rest of `applyArguments`, because we want to have `runArgs.json` as the default values, and fill in, or override as needed.

File: src/mono/wasm/templates/templates/browser/app-support.js [hunk 7]
```diff
@@ -149,44 +147,43 @@ function applyArguments() {
     }
 }
 
-let toAbsoluteUrl = function(possiblyRelativeUrl) { return possiblyRelativeUrl; }
+let toAbsoluteUrl = function (path, prefix) {
```
Comment (radical) [inline src/mono/wasm/templates/templates/browser/app-support.js:150]: this one can be dropped as it gets overwritten below anyway.

File: src/mono/wasm/wasm.proj [hunk 8]
```diff
@@ -210,6 +212,7 @@
 
       <CMakeConfigurationLinkFlags Condition="'$(Configuration)' == 'Debug'"  >$(CMakeConfigurationEmccFlags)</CMakeConfigurationLinkFlags>
       <CMakeConfigurationLinkFlags Condition="'$(Configuration)' == 'Release'">-O2</CMakeConfigurationLinkFlags>
+      <CMakeConfigurationLinkFlags Condition="'$(WasmEnableES6)' != 'false'">$(CMakeConfigurationLinkFlags) -s EXPORT_ES6=1</CMakeConfigurationLinkFlags>
```
Comment (radical) [inline src/mono/wasm/wasm.proj:215]: Why move this from `_EmccLinkFlags` to here?

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs (hunk 2 above)
Comment (radical) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:26]:
```suggestion
    public bool WasmEnableES6 { get; set; }
```

File: src/mono/wasm/templates/templates/browser/app-support.js [hunk 9]
```diff
@@ -149,44 +147,43 @@ function applyArguments() {
...
             if (!Module.config) {
                 const err = new Error("Could not find ./mono-config.json. Cancelling run");
                 set_exit_code(1);
                 throw err;
             }
             // Have to set env vars here to enable setting MONO_LOG_LEVEL etc.
-            for (let variable in runArgs.runtimeArgs) {
-                config.environment_variables[variable] = runArgs.runtimeArgs[variable];
+            for (let variable in runArgs.environment_variables) {
+                config.environment_variables[variable] = runArgs.environment_variables[variable];
```
Comment (radical) [inline src/mono/wasm/templates/templates/browser/app-support.js:186]: `runArgs.environment_variables` -> `runArgs.environmentVariables`
The json has: `"environmentVariables":{`

File: src/mono/wasm/wasm.proj (hunk 8 above)
Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:215]: because if it was bake into `emcc-link.rsp` it could not be un-done on customer machine in workload.

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs (hunk 2 above)
Comment (pavelsavara) [reply src/tasks/WasmAppBuilder/WasmAppBuilder.cs:26]: It didn't work for me., will leave that battle to you :)

**Comment by pavelsavara** [comment]
The `Not found: AOT: image 'async_main_with_args_Release_True' found.` failure is https://github.com/dotnet/runtime/pull/70704
